src/tryGetDebouncedCompletions.ts

```diff
@@ -0,0 +1,61 @@
+import * as vscode from "vscode";
+import { Capability, isCapabilityEnabled } from "./capabilities/capabilities";
+import getInlineCompletionItems from "./getInlineCompletionItems";
+import TabnineInlineCompletionItem from "./inlineSuggestions/tabnineInlineCompletionItem";
+import { sleep, timed } from "./utils/utils";
+
+export default async function tryGetDebouncedCompletions(
+  document: vscode.TextDocument,
+  position: vscode.Position,
+  token: vscode.CancellationToken
+): Promise<
+  vscode.InlineCompletionList<TabnineInlineCompletionItem> | undefined
+> {
+  const { time, value: current } = await timed(() =>
+    getInlineCompletionItems(document, position)
+  );
+
+  const debounceTime = calculateDebounceValue(time);
+
+  if (debounceTime > 0) {
+    await debounceOrCancelOnRequest(token, debounceTime);
+
+    if (token.isCancellationRequested) {
+      return undefined;
+    }
+
+    // re fetch the most updated suggestions
+    return getInlineCompletionItems(document, position);
```

Comment:
I think this is excessive - how about checking the offset (or something similar to JB's `modificationStamp`) before and after sleeping, and re-fetching only if necessary?

Reply:
no, it is irrelevant to the offset,  it is relevant to the binary cache. 
imaging the following use-case: the first request was fired at T0 the result was an empty suggestion because the server did not complete the request,  the sever completed the request at T1 and the cash was filled with the result, if we will re-query the binary after the debounce  there is more chance that we will get the competed  response